proxy/coco/src/announcement.rs

```diff
@@ -0,0 +1,254 @@
+use librad::uri;
+
+use crate::error::Error;
+use crate::oid::Oid;
+use crate::peer;
+
+/// Name for the bucket used in [`kv::Store`].
+const BUCKET_NAME: &str = "announcements";
+/// Key for the single value used as cache.
+const KEY_NAME: &str = "latest";
+
+/// An update and all the required information that can be announced on the network.
+pub type Announcement = (uri::RadUrn, Oid);
+
+/// Announces the list of given `updates` with the [`librad::net::protocol`].
+///
+/// # Errors
+///
+/// * if the announcemnet of one of the project heads failed
+pub async fn announce(
+    protocol: net::protocol::Protocol<net::peer::PeerStorage<keys::SecretKey>, net::peer::Gossip>,
+    updates: impl Iterator<Item = &Announcement> + Send,
+) -> Result<(), Error> {
+    for (urn, hash) in updates {
+        let have = net::peer::Gossip {
+            urn: urn.clone(),
+            rev: Some(net::peer::Rev::Git((*hash).into())),
+            origin: None,
+        };
+
+        protocol.announce(have).await;
+    }
+
+    Ok(())
+}
+
+/// Builds the latest list of [`Announcement`]s for the current state of the peer.
+///
+/// # Errors
+///
+/// * if listing of the projects fails
+/// * if listing of the Refs for a project fails
+pub fn build(api: &peer::Api) -> Result<HashSet<Announcement>, Error> {
+    let mut list: HashSet<Announcement> = HashSet::new();
+
+    // TODO(xla): We need to avoid the case where there is no owner yet for the peer api, there
+    // should be machinery to kick off these routines only if our app state is ready for it.
+    match api.list_projects() {
+        Err(Error::Storage(librad::git::storage::Error::Config(_err))) => Ok(list),
+        Err(err) => Err(err),
+        Ok(projects) => {
+            for project in &projects {
+                let refs = api.list_project_refs(&project.urn())?;
+
+                for (head, hash) in &refs.heads {
+                    list.insert((
+                        uri::RadUrn {
+                            path: head.parse::<uri::Path>()?,
+                            ..project.urn()
+                        },
+                        Oid::from(*hash.deref()),
+                    ));
+                }
+            }
+
+            Ok(list)
+        },
+    }
+}
+
+/// Computes the list of announcements based on the difference of the `new` and `old` state. An
+/// [`Announcement`] will be included if an entry in `new` can't be found in `old`.
+#[allow(clippy::implicit_hasher)]
+#[must_use]
+pub fn diff<'a>(
+    old_state: &'a HashSet<Announcement>,
+    new_state: &'a HashSet<Announcement>,
+) -> HashSet<Announcement> {
+    new_state.difference(old_state).cloned().collect()
+}
+
+/// Load the cached announcements from the [`kv::Store`].
+///
+/// # Errors
+///
+/// * if the [`kv::Bucket`] can't be accessed
+/// * if the access of the key in the [`kv::Bucket`] fails
+#[allow(clippy::or_fun_call)]
+pub fn load(store: &kv::Store) -> Result<HashSet<Announcement>, Error> {
```

Comment:
Ah! I'm just realising that the `kv` is crossing over the boundary to `coco`. I think a suggestion here would be to have a trait for this capability, i.e.

```rust
pub trait AnnouncementStore {
  type Error;
  fn save(&self, updates: HashSet<Announcement>) -> Result<(), Self::Error>;
  fn load(&self) -> Result<HashSet<Announcement>, Self::Error>;
}
```

I'd suggest doing that as a follow-up maybe, but what do you think?

Reply:
#857